### ISSUE TITLE
AtoMMC issues

### ISSUE THREAD

**Issue by hoglet67** [hoglet67/Atomulator#3]
The AtoMMC implementation in Atomulator has some issues:

- the rom image contains  v2.97E - need to update to 3.00E

- *DELETE (on Windows) gives Access Denied. Linux is fine.

- firmware lacks the following commands:
```
#define CMD_DIR_GETCWD		0x03
#define CMD_DIR_MKDIR		0x04
#define CMD_DIR_RMDIR		0x05
#define CMD_RENAME			0x08

```

- RAF test on Windows hits above issue with delete.

- RAF test on Linux hangs Atomulator at:
![Screenshot from 2019-12-19 10-20-25](https://user-images.githubusercontent.com/1951463/71165858-7ccd2e80-2249-11ea-898e-622074804017.png)

**Comment by kwoot** [comment]
Interesting that you can compile this on Linux. Trying on Ubuntu 20.04 but lots of errors. Is there a compiled version somewhere on line?

**Comment by hoglet67** [comment]
If you are seeing errors, it's probably because you don't have the right dependencies installed.

It needs at least there:
- liballegro4-dev
- libopenal-dev
- libalut-dev
- zlib1g-dev

Allegro 4 is getting harder to find on modern distributions.

Also, relating to this thread, I have some work on the dev branch that starts to address some of the AtoMMC issues.

**Comment by hoglet67** [comment]
I've just fixed a couple of issues on Windows:

*DELETE file was failing on Windows because the file was still open, for a couple of reasons:
- the AtoMMC ROM has always opened a file for reading before deleting it (why I don't know). In pulling in the latest AtoMMC code I lost a emulator specific workaround (fixed in 58c0fc5)
- ff_emu was leaking file handles and leaving files open (fixed in c4b6092)

*RMDIR file was failing on Windows because it seems rmdir must be used, rather than unlink (fixed in 2a5f87f)

With these fixes, I can run RAFTEST on Windows.

Here's a list of remaining issues:

(Tested in Atomulator in Windows with AtoMMC 3.00E and the latest dev branch code)

*CWD <dirname> fails silently if dirname is a file, and you end up with a messed up CWD. Subsequent *CAT will then fail. This is an emulation bug; the real AtoMMC fails "NO PATH"

*RMDIR will successfully delete a file, because it calls down to unlink. This not just an emulation issue; the real AtoMMC has the same issue.

*DELETE <dirname> will delete the contents of the directory. This is working as intended, because *DELETE uses the same iterator pattern that *CAT and *INFO use. The real AtoMMC works in the same way.

**Comment by hoglet67** [comment]
All the AtoMMC issues that have been discussed in this thread are now fixed:

58c0fc56 AtoMMC: fix an ACCESS DENIED bug on Windows when deleting files
c4b60929 AtoMMC: really fix the ACCESS DENIED bug
2af587f4 AtoMMC: Fix a bug with RMDIR on Windows
4a675850 AtoMMC: Fix hard crash on reading unopened file on Linux (should be INVALID OBJECT)
be7eb4a4 AtoMMC: Update from 2.97E to 3.01E
e59223a7 AtoMMC: comment out debug logging
7806d1ec AtoMMC: Fix bug that allowed *CWD <file> without an error
bf619bb5 AtoMMC: MKDIR now gives EXISTS error if name already exists
